### DISCUSSION TITLE
Milvus restart resulted in empty Databases

### DISCUSSION THREAD

**Discussion by ganderaj** [milvus-io/milvus#26619] [Q&A and General discussion]
Team,

We have faced an issue with our Milvus deployment on AWS EKS which was unresponsive. Pod logs below.

Proxy:
`[ERROR] [grpcclient/client.go:161] ["failed to get client address"] [error="find no available rootcoord, check rootcoord state"]`

RootCoord:
`[WARN] [client/client.go:94] ["QueryCoordClient msess key not existed"] [key=querycoord]`
`[ERROR] [grpcclient/client.go:161] ["failed to get client address"] [error="find no available querycoord, check querycoord state"]`
`[WARN] [client/client.go:98] ["DataCoordClient, not existed in msess "] [key=datacoord] ["len of msess"=0]`
`[ERROR] [grpcclient/client.go:161] ["failed to get client address"] [error="find no available datacoord, check datacoord state"] `

DataCoord:
`[ERROR] [grpcclient/client.go:161] ["failed to get client address"] [error="find no available rootcoord, check rootcoord state"]`

In an attempt to resurrect environment, we have restarted all pods (took backup of ETCD PVCs in prior) which helped us. However, the DB on Milvus lost all its collections and even Roles / Grants which were applied earlier.

We confirm there was no changes to the ETCD configurations; ETCD pods were restarted but the underlying PVCs have been repurposed. This is confirmed by inspecting the data on PVCs and even age of PVCs.

Is there a way restore lost collections on our DBs ? If there is none, what is a recommended DR strategy for future incidents ?

Please be noted below are the pod logs taken in real-time before restarting. There could be other errors reported by other coords / nodes.

[DataCoord.log](https://github.com/milvus-io/milvus/files/12447061/DataCoord.log)
[Proxy.log](https://github.com/milvus-io/milvus/files/12447062/Proxy.log)
[RootCoord.log](https://github.com/milvus-io/milvus/files/12447063/RootCoord.log)

**Comment by ganderaj** [comment]
@haorenfsa / @xiaofan-luan 

Post Milvus restart our DBs have turned empty despite repurposing the same ETCD PVCs. Can you please look into this and revert if there is a way we can retrieve lost data ? More details in the post above.

**Comment by xiaofan-luan** [comment]
> 

From your description this is pretty sure all your etcd data has been lost.
You can use etcdctl or birdwatcher to read the etcd and check if there are anything left(My suspect is that there are something wrong with the way you operate etcd).

If you are sure you etcd data is backuped you can try to do the same operation to another dir and check if the data is there. with the right etcd data your data can be back.

**Reply by xiaofan-luan** [reply]
But I highly recommend not to play with etcds unless you are very familiar about milvus.
etcd, S3 and kafka/pulsar(optional if you don't worry about recent data loss) need to match each other

**Reply by ganderaj** [reply]
This is a disappointing news. What could potentially lead to data loss during Milvus restart ?

This is the second time we lost data post a restart. The first time when it happened, our Milvus behaved as if it is a fresh new installation; all DBs, Roles and Users were lost. In fact even our root account password was reset to default "Milvus".

This second incident is a bit peculiar; Grants for Roles have been reset but not the roles. DBs are present but not the collections within.

Can you please suggest best practices in deploying Milvus ? Especially, what can we do to ensure data is not lost in any cases, a DR Strategy ?

**Reply by ganderaj** [reply]
_"If you are sure you etcd data is backuped you can try to do the same operation to another dir and check if the data is there. with the right etcd data your data can be back."_

Yes, I have the PVC data archived to a local disk. How do I make use of this data ? Please guide with steps/doc.

**Comment by xiaofan-luan** [comment]
> Can you please suggest best practices in deploying Milvus ? Especially, what can we do to ensure data is not lost in any cases, a DR Strategy ?

Milvus should never lose data when restart. 
the thing is what happened to your etcd?

**Comment by xiaofan-luan** [comment]
This is what you should investigate on "In an attempt to resurrect environment, we have restarted all pods (took backup of ETCD PVCs in prior) which helped us. However, the DB on Milvus lost all its collections and even Roles / Grants which were applied earlier.

We confirm there was no changes to the ETCD configurations; ETCD pods were restarted but the underlying PVCs have been repurposed. This is confirmed by inspecting the data on PVCs and even age of PVCs."

**Comment by xiaofan-luan** [comment]
etcd persist your data and could you check why etcd data is missing?

**Reply by ganderaj** [reply]
ETCD data is not missing. PVCs have been repurposed and it holds the data just that Milvus says otherwise; DBs are empty.

**Reply by xiaofan-luan** [reply]
did you check with etcdcli?
I'm pretty sure the etcd data is missing

**Reply by ganderaj** [reply]
How do I check with etcdcli ? Any command in precise that I can try to validate whether data exists ?

**Comment by LoveEachDay** [comment]
@ganderaj Could you provide a detailed steps for deploying your milvus cluster on AWS EKS? We'll not recreate any pvc when restarting etcd pods.

**Reply by ganderaj** [reply]
What we have done with implementing Milvus Cluster on AWS EKS.

 1. Pulled helm chart to our local machine from [Milvus Helm Chart](https://milvus-io.github.io/milvus-helm/)
 2. Made changes to `values.yaml` files to meet our expectations. Environment details redacted; [values_redacted.txt](https://github.com/milvus-io/milvus/files/12459318/values_redacted.txt)
 3. Deployed Milvus using the command -- `helm install milvus ./milvus -n milvus`

As mentioned over my initial post, our environment turned unresponsive with each component blaming other component unreachable. We have ran `helm upgrade –install milvus ./milvus -n milvus  --recreate-pods ` which restarted our environment making it reachable.

Unfortunately, all our collections within the databases are lost. We are currently looking for possibility of restoring these collections considering ETCD PVCs still exist.
